File: companion-ios/Sources/H3Tiles.swift
import Foundation
import CoreLocation

// One H3 res-6 map tile: its H3 id (the device's filename), the raw cell, and
// lat/lng bbox.
struct MapTile: Identifiable, Equatable {
    let id: String            // H3 hex id, e.g. "86283082fffffff"
    let cell: UInt64          // raw H3 index (for the hexagon boundary)
    let south: Double
    let west: Double
    let north: Double
    let east: Double

    // The tile's actual hexagon outline (not its bounding box).
    var hexagon: [CLLocationCoordinate2D] { H3Tiles.hexagon(of: cell) }
    // Center of the hexagon (for placing the on-device check mark).
    var center: CLLocationCoordinate2D {
        .init(latitude: (south + north) / 2, longitude: (west + east) / 2)
    }
}

// Thin Swift front for the H3 C shim (Sources/H3/h3shim.*). Res-6 hexagons are
// ~5.6 km across; the whole scheme is deterministic and global, so the same
// ground always maps to the same tile id — that's what makes dedup work.
enum H3Tiles {
    // Every res-6 tile overlapping the drawn box, in a stable order.
    static func coveringTiles(south s: Double, west w: Double,
                              north n: Double, east e: Double) -> [MapTile] {
        var ids = [UInt64](repeating: 0, count: 4096)
        let count = Int(ids.withUnsafeMutableBufferPointer {
            h3_covering_cells(s, w, n, e, $0.baseAddress, Int32($0.count))
        })
        return (0..<count).map { i in tile(from: ids[i]) }
    }

    static func tile(from cell: UInt64) -> MapTile {
        var ts = 0.0, tw = 0.0, tn = 0.0, te = 0.0
        h3_cell_bbox(cell, &ts, &tw, &tn, &te)
        var idBuf = [CChar](repeating: 0, count: 17)
        h3_cell_id(cell, &idBuf, 17)
        return MapTile(id: String(cString: idBuf), cell: cell,
                       south: ts, west: tw, north: tn, east: te)
    }

    // The res-6 H3 id containing a coordinate (for hit-testing map taps).
    static func id(at coord: CLLocationCoordinate2D) -> String? {
        let cell = h3_cell_at(coord.latitude, coord.longitude)
        guard cell != 0 else { return nil }
        var idBuf = [CChar](repeating: 0, count: 17)
        h3_cell_id(cell, &idBuf, 17)
        return String(cString: idBuf)
    }

    // The cell's hexagon outline as map coordinates.
    static func hexagon(of cell: UInt64) -> [CLLocationCoordinate2D] {
        var buf = [Double](repeating: 0, count: 24)   // up to 12 verts × 2
        let n = Int(buf.withUnsafeMutableBufferPointer {
            h3_cell_boundary(cell, $0.baseAddress, Int32($0.count / 2))
        })
        return (0..<n).map { CLLocationCoordinate2D(latitude: buf[$0 * 2], longitude: buf[$0 * 2 + 1]) }
    }
}

File: companion-android/app/src/main/java/com/raemond/opentrailpaper/map/H3Tiles.kt
package com.raemond.opentrailpaper.map

import com.raemond.opentrailpaper.data.LatLon

/** Raw JNI surface over the vendored H3 C (see cpp/h3jni.c). */
internal object H3Native {
    init {
        System.loadLibrary("h3shim")
    }

    external fun coveringCells(south: Double, west: Double, north: Double, east: Double): LongArray
    external fun cellBbox(cell: Long): DoubleArray
    external fun cellBoundary(cell: Long): DoubleArray
    external fun cellId(cell: Long): String
    external fun fromId(id: String): Long
    external fun cellAt(lat: Double, lng: Double): Long
}

/**
 * One H3 res-6 map tile: its H3 id (the device's filename), the raw cell, and
 * its lat/lng bbox.
 */
data class MapTile(
    val id: String,          // H3 hex id, e.g. "86283082fffffff"
    val cell: Long,          // raw H3 index (for the hexagon boundary)
    val south: Double,
    val west: Double,
    val north: Double,
    val east: Double,
) {
    /** The tile's actual hexagon outline (not its bounding box). */
    val hexagon: List<LatLon> by lazy { H3Tiles.hexagon(cell) }

    /** Centre of the hexagon (for placing the on-device check mark). */
    val center: LatLon get() = LatLon((south + north) / 2, (west + east) / 2)
}

/**
 * Thin front for the H3 C shim. Res-6 hexagons are ~5.6 km across; the whole
 * scheme is deterministic and global, so the same ground always maps to the same
 * tile id — that's what makes dedup with the device (and with the iOS app) work.
 */
object H3Tiles {

    /** Every res-6 tile overlapping the drawn box, in a stable order. */
    fun coveringTiles(south: Double, west: Double, north: Double, east: Double): List<MapTile> =
        H3Native.coveringCells(south, west, north, east).map { tile(it) }

    fun tile(cell: Long): MapTile {
        val b = H3Native.cellBbox(cell)
        return MapTile(H3Native.cellId(cell), cell, b[0], b[1], b[2], b[3])
    }

    /**
     * Geometry for an id we already know (a cell on the device, a gap in
     * coverage) — the inverse of [idAt]. null if the string isn't a cell.
     */
    fun tile(id: String): MapTile? {
        val cell = H3Native.fromId(id)
        return if (cell == 0L) null else tile(cell)
    }

    /** The res-6 H3 id containing a coordinate (for hit-testing map taps). */
    fun idAt(coord: LatLon): String? {
        val cell = H3Native.cellAt(coord.lat, coord.lon)
        return if (cell == 0L) null else H3Native.cellId(cell)
    }

    /** The cell's hexagon outline as map coordinates. */
    fun hexagon(cell: Long): List<LatLon> {
        val flat = H3Native.cellBoundary(cell)
        return (0 until flat.size / 2).map { LatLon(flat[it * 2], flat[it * 2 + 1]) }
    }
}
